client/consensus/pow/src/lib.rs

```diff
@@ -243,55 +249,165 @@ impl<B: BlockT, C, S, Algorithm> PowVerifier<B, C, S, Algorithm> {
 				inherent_data,
 				timestamp_now
 			)?;
 
-			body = Some(block.deconstruct().1);
+			block.body = Some(check_block.deconstruct().1);
 		}
 
-		let key = aux_key(&hash);
+		let inner_seal = match block.post_digests.last() {
+			Some(DigestItem::Seal(id, seal)) => {
+				if id == &POW_ENGINE_ID {
+					seal.clone()
+				} else {
+					return Err(Error::<B>::WrongEngine(*id).into())
+				}
+			},
+			_ => return Err(Error::<B>::HeaderUnsealed(block.header.hash()).into()),
+		};
+
+		let intermediate = PowIntermediate::<B, Algorithm::Difficulty>::decode(
+			&mut &block.intermediates.remove(INTERMEDIATE_KEY)
+				.ok_or(Error::<B>::NoIntermediate)?[..]
+		).map_err(|_| Error::<B>::NoIntermediate)?;
+
+		let difficulty = match intermediate.difficulty {
+			Some(difficulty) => difficulty,
+			None => self.algorithm.difficulty(&BlockId::hash(parent_hash))?,
+		};
+
+		if !self.algorithm.verify_difficulty(
+			difficulty,
+			&BlockId::hash(parent_hash),
+			&inner_seal,
+		)? {
+			return Err(Error::<B>::InvalidDifficulty.into())
+		}
+
+		aux.difficulty = difficulty;
+		aux.total_difficulty.increment(difficulty);
+
+		let key = aux_key(&intermediate.hash);
+		block.auxiliary.push((key, Some(aux.encode())));
+		if block.fork_choice.is_none() {
+			block.fork_choice = Some(ForkChoiceStrategy::Custom(
+				aux.total_difficulty > best_aux.total_difficulty
+			));
+		}
+
+		self.inner.import_block(block, new_cache).map_err(Into::into)
+	}
+}
+
+/// A verifier for PoW blocks.
+pub struct PowVerifier<B: BlockT, Algorithm> {
+	algorithm: Algorithm,
+	_marker: PhantomData<B>,
+}
+
+impl<B: BlockT, Algorithm> PowVerifier<B, Algorithm> {
+	pub fn new(
+		algorithm: Algorithm,
+	) -> Self {
+		Self { algorithm, _marker: PhantomData }
+	}
+
+	fn check_header(
+		&self,
+		mut header: B::Header,
+	) -> Result<(B::Header, DigestItem<B::Hash>), Error<B>> where
+		Algorithm: PowAlgorithm<B>,
+	{
+		let hash = header.hash();
+
+		let (seal, inner_seal) = match header.digest_mut().pop() {
+			Some(DigestItem::Seal(id, seal)) => {
+				if id == POW_ENGINE_ID {
+					(DigestItem::Seal(id, seal.clone()), seal)
+				} else {
+					return Err(Error::WrongEngine(id))
+				}
+			},
+			_ => return Err(Error::HeaderUnsealed(hash)),
+		};
+
+		let pre_hash = header.hash();
+
+		if !self.algorithm.verify_seal(
+			&pre_hash,
+			&inner_seal,
+		)? {
+			return Err(Error::InvalidSeal);
+		}
+
+		Ok((header, seal))
+	}
+}
+
+impl<B: BlockT, Algorithm> Verifier<B> for PowVerifier<B, Algorithm> where
+	Algorithm: PowAlgorithm<B> + Send + Sync,
+{
+	fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justification: Option<Justification>,
+		body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B, ()>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let hash = header.hash();
+		let (checked_header, seal) = self.check_header(header)?;
+
+		let intermediate = PowIntermediate::<B, Algorithm::Difficulty> {
+			hash: hash,
+			difficulty: None,
```

Comment:
should we not verify the difficulty here in the `Verifier`, as that's the parallelizable operation?

Reply:
Difficulty may not be parallelizable operation. The mining algorithm is always fixed (or can only be chosen from a set of validated ones), but for difficulty adjustment algorithms, they can be defined in runtime. In those cases, block importer will have to do a runtime call on the parent block to fetch the difficulty. Because parent block may not be available in `Verifier`, we can only set it to `None` here, meaning we do not yet know the value.

However, in mining routine, the value of difficulty is indeed available because we mine on local best block. In that case, difficulty is set to `Some` so we can avoid double calculation.

Reply:
Difficulty may not be parallelizable operation. The mining algorithm is always fixed (or can only be chosen from a set of validated ones), but for difficulty adjustment algorithms, they can be defined in runtime. In those cases, block importer will have to do a runtime call on the parent block to fetch the difficulty. Because parent block may not be available in `Verifier`, we can only set it to `None` here, meaning we do not yet know the value.

However, in mining routine, the value of difficulty is indeed available because we mine on local best block. In that case, difficulty is set to `Some` so we can avoid double calculation.